File: FourSpacesFinder-MacOS/Sources/Math.swift
import Foundation

// MARK: - Fraction

struct Fraction: Equatable, CustomStringConvertible, Hashable {
    let numerator: Int
    let denominator: Int
    
    init(_ n: Int, _ d: Int = 1) {
        if d == 0 { fatalError("Denominator cannot be zero") }
        let common = Fraction.gcd(abs(n), abs(d))
        let sign = (n < 0) == (d < 0) ? 1 : -1
        self.numerator = (abs(n) / common) * sign
        self.denominator = abs(d) / common
    }
    
    private static func gcd(_ a: Int, _ b: Int) -> Int {
        return b == 0 ? a : gcd(b, a % b)
    }
    
    var description: String {
        return denominator == 1 ? "\(numerator)" : "\(numerator)/\(denominator)"
    }
    
    static let zero = Fraction(0)
    static let one = Fraction(1)
    
    static func parse(_ raw: String) -> Fraction? {
        let s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if s.isEmpty { return nil }
        
        let parts = s.components(separatedBy: "/")
        if parts.count == 1 {
            guard let n = Int(parts[0]) else { return nil }
            return Fraction(n)
        }
        
        if parts.count == 2 {
            guard let n = Int(parts[0]), let d = Int(parts[1]), d != 0 else { return nil }
            return Fraction(n, d)
        }
        
        return nil
    }
    
    func add(_ rhs: Fraction) -> Fraction {
        return Fraction(
            numerator * rhs.denominator + rhs.numerator * denominator,
            denominator * rhs.denominator
        )
    }
    
    func sub(_ rhs: Fraction) -> Fraction {
        return Fraction(
            numerator * rhs.denominator - rhs.numerator * denominator,
            denominator * rhs.denominator
        )
    }
    
    func mul(_ rhs: Fraction) -> Fraction {
        return Fraction(
            numerator * rhs.numerator,
            denominator * rhs.denominator
        )
    }
    
    func div(_ rhs: Fraction) -> Fraction {
        return Fraction(
            numerator * rhs.denominator,
            denominator * rhs.numerator
        )
    }
    
    func negate() -> Fraction {
        return Fraction(-numerator, denominator)
    }
    
    var doubleValue: Double {
        return Double(numerator) / Double(denominator)
    }
}

// MARK: - Matrix

struct CalculationStep: Identifiable {
    let id: UUID
    let matrixState: Matrix
    let description: String
    let isPivotStep: Bool
}

struct Matrix: Equatable {
    let rows: Int
    let cols: Int
    var data: [[Fraction]]
    
    init(rows: Int, cols: Int, fill: Fraction = .zero) {
        self.rows = rows
        self.cols = cols
        self.data = Array(repeating: Array(repeating: fill, count: cols), count: rows)
    }
    
    func get(_ r: Int, _ c: Int) -> Fraction {
        return data[r][c]
    }
    
    mutating func set(_ r: Int, _ c: Int, _ val: Fraction) {
        data[r][c] = val
    }
    
    func transposed() -> Matrix {
        var res = Matrix(rows: cols, cols: rows)
        for r in 0..<rows {
            for c in 0..<cols {
                res.set(c, r, self.get(r, c))
            }
        }
        return res
    }
    
    func rrefWithSteps() -> (rrefMatrix: Matrix, pivots: [Int], steps: [CalculationStep]) {
        var m = self // Value type copy
        var pivotRow = 0
        var pivotIndices: [Int] = []
        var steps: [CalculationStep] = []
        
        steps.append(CalculationStep(id: UUID(), matrixState: m, description: "", isPivotStep: false))
        
        for col in 0..<cols {
            if pivotRow >= rows { break }
            
            var pivot = pivotRow
            while pivot < rows && m.get(pivot, col) == .zero {
                pivot += 1
            }
            
            if pivot < rows {
                if pivot != pivotRow {
                    let temp = m.data[pivotRow]
                    m.data[pivotRow] = m.data[pivot]
                    m.data[pivot] = temp
                    
                    steps.append(CalculationStep(id: UUID(), matrixState: m, description: "E\(pivot + 1)\(pivotRow + 1)", isPivotStep: false))
                }
                
                let divisor = m.get(pivotRow, col)
                if divisor != .one {
                    for c in 0..<cols {
                        m.set(pivotRow, c, m.get(pivotRow, c).div(divisor))
                    }
                    steps.append(CalculationStep(id: UUID(), matrixState: m, description: "E\(pivotRow + 1)(1/\(divisor))", isPivotStep: false))
                }
                
                for r in 0..<rows {
                    if r != pivotRow {
                        let factor = m.get(r, col)
                        if factor != .zero {
                            for c in 0..<cols {
                                m.set(r, c, m.get(r, c).sub(factor.mul(m.get(pivotRow, c))))
                            }
                            steps.append(CalculationStep(id: UUID(), matrixState: m, description: "E\(r + 1)\(pivotRow + 1)(\(factor.negate()))", isPivotStep: false))
                        }
                    }
                }
                
                pivotIndices.append(col)
                pivotRow += 1
            }
        }
        
        steps.append(CalculationStep(id: UUID(), matrixState: m, description: "", isPivotStep: true))
        
        return (m, pivotIndices, steps)
    }
    
    func columnSpaceBasis(pivots: [Int]) -> [String] {
        var basis: [String] = []
        for colIndex in pivots {
            var vecStr = ""
            for r in 0..<rows {
                vecStr += "[\(self.get(r, colIndex))]\n"
            }
            basis.append(vecStr)
        }
        return basis
    }
    
    func nullSpaceBasis(rrefM: Matrix, pivots: [Int]) -> [String] {
        let pivotSet = Set(pivots)
        var basis: [String] = []
        
        for c in 0..<cols {
            if !pivotSet.contains(c) {
                var sol = Array(repeating: Fraction.zero, count: cols)
                sol[c] = .one
                
                for (rowIndex, pivotCol) in pivots.enumerated() {
                    sol[pivotCol] = rrefM.get(rowIndex, c).negate()
                }
                
                var vecStr = ""
                for val in sol {
                    vecStr += "[\(val)]\n"
                }
                basis.append(vecStr)
            }
        }
        
        if basis.isEmpty { return ["[ 0 ]\n(trivial)"] }
        return basis
    }
    
    func rowSpaceBasisFromRref(rrefM: Matrix) -> [String] {
        var basis: [String] = []
        
        for r in 0..<rrefM.rows {
            var allZero = true
            for c in 0..<rrefM.cols {
                if rrefM.get(r, c) != .zero {
                    allZero = false
                    break
                }
            }
            if allZero { continue }
            
            var vecStr = ""
            for c in 0..<rrefM.cols {
                vecStr += "[\(rrefM.get(r, c))]\n"
            }
            basis.append(vecStr)
        }
        
        if basis.isEmpty { return ["[ 0 ]\n(trivial)"] }
        return basis
    }
}

// MARK: - Helpers

func parseVectorDoubles(raw: String, dimension: Int) -> [Double] {
    let lines = raw.components(separatedBy: "\n").map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }.filter { !$0.isEmpty }
    var values: [String] = []
    
    for line in lines {
        if line.hasPrefix("(") { continue }
        if line.hasPrefix("[") && line.hasSuffix("]") {
            let inner = String(line.dropFirst().dropLast()).trimmingCharacters(in: .whitespacesAndNewlines)
            if !inner.isEmpty { values.append(inner) }
            continue
        }
        if !line.isEmpty { values.append(line) }
    }
    
    var doubles = values.map { parseScalarToDouble($0) ?? 0 }
    if doubles.isEmpty { doubles = [0] }
    
    if doubles.count < dimension {
        doubles += Array(repeating: 0.0, count: dimension - doubles.count)
    } else if doubles.count > dimension {
        doubles = Array(doubles.prefix(dimension))
    }
    return doubles
}

func fractionToLatex(_ f: Fraction) -> String {
    if f.denominator == 1 { return "\(f.numerator)" }
    if f.numerator < 0 { return "-\\frac{\(abs(f.numerator))}{\(f.denominator)}" }
    return "\\frac{\(f.numerator)}{\(f.denominator)}"
}

func matrixToLatex(_ matrix: Matrix) -> String {
    let rows = (0..<matrix.rows).map { r in
        (0..<matrix.cols).map { c in
            fractionToLatex(matrix.get(r, c))
        }.joined(separator: " & ")
    }.joined(separator: " \\\\ ")
    return "\\begin{bmatrix} \(rows) \\end{bmatrix}"
}

private func scalarStringToLatex(_ raw: String) -> String {
    let s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    let parts = s.split(separator: "/").map(String.init)
    if parts.count == 2, let n = Int(parts[0]), let d = Int(parts[1]), d != 0 {
        if n < 0 { return "-\\frac{\(abs(n))}{\(d)}" }
        return "\\frac{\(n)}{\(d)}"
    }
    return s.isEmpty ? "0" : s
}

func vectorStringToLatex(_ vectorString: String) -> String {
    let lines = vectorString
        .components(separatedBy: "\n")
        .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
        .filter { !$0.isEmpty }
    
    var values: [String] = []
    for line in lines {
        if line.hasPrefix("(") { continue }
        if line.hasPrefix("[") && line.hasSuffix("]") {
            let inner = String(line.dropFirst().dropLast()).trimmingCharacters(in: .whitespacesAndNewlines)
            if !inner.isEmpty { values.append(inner) }
            continue
        }
        values.append(line)
    }
    
    if values.isEmpty { values.append("0") }
    let body = values.map(scalarStringToLatex).joined(separator: " \\\\ ")
    return "\\begin{bmatrix} \(body) \\end{bmatrix}"
}

func parseScalarToDouble(_ raw: String) -> Double? {
    let s = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    if s.isEmpty { return nil }
    let parts = s.components(separatedBy: "/")
    if parts.count == 2 {
        guard let n = Double(parts[0]), let d = Double(parts[1]), d != 0 else { return nil }
        return n / d
    }
    return Double(s)
}

func isZeroVector(_ v: [Double], eps: Double = 1e-9) -> Bool {
    let len = sqrt(v.reduce(0) { $0 + $1 * $1 })
    return len < eps
}

func cross2(_ a: [Double], _ b: [Double]) -> Double {
    return (a.count > 1 && b.count > 1) ? (a[0] * b[1] - a[1] * b[0]) : 0
}

File: FourSpacesFinder-Web/lib/math.ts
import { v4 as uuidv4 } from 'uuid';

// MARK: - Fraction

export class Fraction {
  numerator: number;
  denominator: number;

  constructor(n: number, d: number = 1) {
    if (d === 0) throw new Error("Denominator cannot be zero");
    const common = this.gcd(Math.abs(n), Math.abs(d));
    const sign = (n < 0) === (d < 0) ? 1 : -1;
    this.numerator = (Math.abs(n) / common) * sign;
    this.denominator = Math.abs(d) / common;
  }

  private gcd(a: number, b: number): number {
    return b === 0 ? a : this.gcd(b, a % b);
  }

  toString(): string {
    return this.denominator === 1 ? `${this.numerator}` : `${this.numerator}/${this.denominator}`;
  }

  static zero = new Fraction(0);
  static one = new Fraction(1);
  
  static parse(raw: string): Fraction | null {
    const s = raw.trim();
    if (s.length === 0) return null;
    
    const parts = s.split('/');
    if (parts.length === 1) {
      const n = parseInt(parts[0], 10);
      if (Number.isNaN(n)) return null;
      return new Fraction(n);
    }
    
    if (parts.length === 2) {
      const n = parseInt(parts[0], 10);
      const d = parseInt(parts[1], 10);
      if (Number.isNaN(n) || Number.isNaN(d) || d === 0) return null;
      return new Fraction(n, d);
    }
    
    return null;
  }

  add(rhs: Fraction): Fraction {
    return new Fraction(
      this.numerator * rhs.denominator + rhs.numerator * this.denominator,
      this.denominator * rhs.denominator
    );
  }

  sub(rhs: Fraction): Fraction {
    return new Fraction(
      this.numerator * rhs.denominator - rhs.numerator * this.denominator,
      this.denominator * rhs.denominator
    );
  }

  mul(rhs: Fraction): Fraction {
    return new Fraction(
      this.numerator * rhs.numerator,
      this.denominator * rhs.denominator
    );
  }

  div(rhs: Fraction): Fraction {
    return new Fraction(
      this.numerator * rhs.denominator,
      this.denominator * rhs.numerator
    );
  }

  negate(): Fraction {
    return new Fraction(-this.numerator, this.denominator);
  }

  equals(rhs: Fraction): boolean {
    return this.numerator === rhs.numerator && this.denominator === rhs.denominator;
  }

  toNumber(): number {
    return this.numerator / this.denominator;
  }
}

// MARK: - Matrix

export interface CalculationStep {
  id: string;
  matrixState: Matrix; // We'll store a copy
  description: string;
  isPivotStep: boolean;
}

export class Matrix {
  rows: number;
  cols: number;
  data: Fraction[][];

  constructor(rows: number, cols: number, fill: Fraction = Fraction.zero) {
    this.rows = rows;
    this.cols = cols;
    this.data = Array(rows).fill(null).map(() => Array(cols).fill(fill));
  }

  clone(): Matrix {
    const m = new Matrix(this.rows, this.cols);
    for (let r = 0; r < this.rows; r++) {
      for (let c = 0; c < this.cols; c++) {
        m.data[r][c] = this.data[r][c];
      }
    }
    return m;
  }

  get(r: number, c: number): Fraction {
    return this.data[r][c];
  }

  set(r: number, c: number, val: Fraction) {
    this.data[r][c] = val;
  }

  transposed(): Matrix {
    const res = new Matrix(this.cols, this.rows);
    for (let r = 0; r < this.rows; r++) {
      for (let c = 0; c < this.cols; c++) {
        res.set(c, r, this.get(r, c));
      }
    }
    return res;
  }

  rrefWithSteps(): { rrefMatrix: Matrix; pivots: number[]; steps: CalculationStep[] } {
    const m = this.clone();
    let pivotRow = 0;
    const pivotIndices: number[] = [];
    const steps: CalculationStep[] = [];

    steps.push({
      id: uuidv4(),
      matrixState: m.clone(),
      description: "",
      isPivotStep: false,
    });

    for (let col = 0; col < this.cols; col++) {
      if (pivotRow >= this.rows) break;

      let pivot = pivotRow;
      while (pivot < this.rows && m.get(pivot, col).equals(Fraction.zero)) {
        pivot++;
      }

      if (pivot < this.rows) {
        if (pivot !== pivotRow) {
          // Swap rows
          const temp = m.data[pivotRow];
          m.data[pivotRow] = m.data[pivot];
          m.data[pivot] = temp;

          steps.push({
            id: uuidv4(),
            matrixState: m.clone(),
            description: `E${pivot + 1}${pivotRow + 1}`,
            isPivotStep: false,
          });
        }

        const divisor = m.get(pivotRow, col);
        if (!divisor.equals(Fraction.one)) {
          for (let c = 0; c < this.cols; c++) {
            m.set(pivotRow, c, m.get(pivotRow, c).div(divisor));
          }
          steps.push({
            id: uuidv4(),
            matrixState: m.clone(),
            description: `E${pivotRow + 1}(1/${divisor})`,
            isPivotStep: false,
          });
        }

        for (let r = 0; r < this.rows; r++) {
          if (r !== pivotRow) {
            const factor = m.get(r, col);
            if (!factor.equals(Fraction.zero)) {
              for (let c = 0; c < this.cols; c++) {
                // m[r, c] = m[r, c] - (factor * m[pivotRow, c])
                m.set(r, c, m.get(r, c).sub(factor.mul(m.get(pivotRow, c))));
              }
              steps.push({
                id: uuidv4(),
                matrixState: m.clone(),
                description: `E${r + 1}${pivotRow + 1}(${factor.negate()})`,
                isPivotStep: false,
              });
            }
          }
        }

        pivotIndices.push(col);
        pivotRow++;
      }
    }

    steps.push({
      id: uuidv4(),
      matrixState: m.clone(),
      description: "",
      isPivotStep: true,
    });

    return { rrefMatrix: m, pivots: pivotIndices, steps };
  }

  columnSpaceBasis(pivots: number[]): string[] {
    const basis: string[] = [];
    for (const colIndex of pivots) {
      let vecStr = "";
      for (let r = 0; r < this.rows; r++) {
        vecStr += `[${this.get(r, colIndex).toString()}]\n`;
      }
      basis.push(vecStr);
    }
    return basis;
  }

  nullSpaceBasis(rrefM: Matrix, pivots: number[]): string[] {
    const pivotSet = new Set(pivots);
    const basis: string[] = [];

    for (let c = 0; c < this.cols; c++) {
      if (!pivotSet.has(c)) {
        const sol = Array(this.cols).fill(Fraction.zero);
        sol[c] = Fraction.one;

        // The pivots array gives us (rowIndex, pivotCol) pairs essentially
        // pivots[i] is the column index of the pivot in row i
        for (let rowIndex = 0; rowIndex < pivots.length; rowIndex++) {
          const pivotCol = pivots[rowIndex];
          // sol[pivotCol] = -(rrefM[rowIndex, c])
          sol[pivotCol] = rrefM.get(rowIndex, c).negate();
        }

        let vecStr = "";
        for (const val of sol) {
          vecStr += `[${val.toString()}]\n`;
        }
        basis.push(vecStr);
      }
    }

    if (basis.length === 0) return ["[ 0 ]\n(trivial)"];
    return basis;
  }

  rowSpaceBasisFromRref(rrefM: Matrix): string[] {
    const basis: string[] = [];

    for (let r = 0; r < rrefM.rows; r++) {
      let allZero = true;
      for (let c = 0; c < rrefM.cols; c++) {
        if (!rrefM.get(r, c).equals(Fraction.zero)) {
          allZero = false;
          break;
        }
      }
      if (allZero) continue;

      let vecStr = "";
      for (let c = 0; c < rrefM.cols; c++) {
        vecStr += `[${rrefM.get(r, c).toString()}]\n`;
      }
      basis.push(vecStr);
    }

    if (basis.length === 0) return ["[ 0 ]\n(trivial)"];
    return basis;
  }
}

// MARK: - Helpers

function scalarStringToLatex(raw: string): string {
  const s = raw.trim();
  const m = s.match(/^(-?\d+)\/(\d+)$/);
  if (m) {
    const n = parseInt(m[1], 10);
    const d = m[2];
    if (n < 0) return `-\\frac{${Math.abs(n)}}{${d}}`;
    return `\\frac{${n}}{${d}}`;
  }
  return s.length === 0 ? "0" : s;
}

export function fractionToLatex(f: Fraction): string {
  if (f.denominator === 1) return `${f.numerator}`;
  if (f.numerator < 0) return `-\\frac{${Math.abs(f.numerator)}}{${f.denominator}}`;
  return `\\frac{${f.numerator}}{${f.denominator}}`;
}

export function matrixToLatex(matrix: Matrix): string {
  const rows = matrix.data.map(row => row.map(fractionToLatex).join(' & ')).join(' \\\\ ');
  return `\\begin{bmatrix} ${rows} \\end{bmatrix}`;
}

export function vectorStringToLatex(vectorString: string): string {
  const lines = vectorString
    .split('\n')
    .map(s => s.trim())
    .filter(s => s.length > 0);

  const values: string[] = [];
  for (const line of lines) {
    if (line.startsWith('(')) continue;
    if (line.startsWith('[') && line.endsWith(']')) {
      const inner = line.slice(1, -1).trim();
      if (inner.length > 0) values.push(inner);
      continue;
    }
    if (line.length > 0) values.push(line);
  }

  if (values.length === 0) values.push("0");
  const body = values.map(scalarStringToLatex).join(' \\\\ ');
  return `\\begin{bmatrix} ${body} \\end{bmatrix}`;
}

export function parseVectorDoubles(raw: string, dimension: number): number[] {
  const lines = raw.split('\n').map(s => s.trim()).filter(s => s.length > 0);
  const values: string[] = [];
  
  for (const line of lines) {
    if (line.startsWith('(')) continue;
    if (line.startsWith('[') && line.endsWith(']')) {
      const inner = line.slice(1, -1).trim();
      if (inner.length > 0) values.push(inner);
      continue;
    }
    if (line.length > 0) values.push(line);
  }
  
  let doubles = values.map(s => parseScalarToDouble(s) ?? 0);
  if (doubles.length === 0) doubles = [0];
  
  if (doubles.length < dimension) {
    doubles = [...doubles, ...Array(dimension - doubles.length).fill(0)];
  } else if (doubles.length > dimension) {
    doubles = doubles.slice(0, dimension);
  }
  return doubles;
}

function parseScalarToDouble(raw: string): number | null {
  const s = raw.trim();
  if (s.length === 0) return null;
  const parts = s.split('/');
  if (parts.length === 2) {
    const n = parseFloat(parts[0]);
    const d = parseFloat(parts[1]);
    if (!isNaN(n) && !isNaN(d) && d !== 0) {
      return n / d;
    }
  }
  const val = parseFloat(s);
  return isNaN(val) ? null : val;
}

export function isZeroVector(v: number[], eps: number = 1e-9): boolean {
  const len = Math.sqrt(v.reduce((acc, val) => acc + val * val, 0));
  return len < eps;
}

export function cross2(a: number[], b: number[]): number {
    return (a.length > 1 && b.length > 1) ? (a[0] * b[1] - a[1] * b[0]) : 0;
}
